### PR TITLE
Comments

### PR DESCRIPTION
**Pull request by mbrandenburger**
Thanks for this nice first draft. I added a bunch of comments and propose additional content.

### CODE REVIEW

**Comment by samuelvenzi** [comment]
Hey, I have no further comments. I double down on Marcus' points.

File: README.md [hunk 1]
```diff
@@ -1,83 +1,299 @@
 # FPC and CC-tools integration
 
-This guide shows how to deploy and run a chaincode using cc-tools on an FPC network ensuring the confidentiality requirements achieved by [fabric-private-chaincode](https://github.com/hyperledger/fabric-private-chaincode).
+The goal of this project is to combine the strength of [cc-tools](https://github.com/hyperledger-labs/cc-tools) with the enhanced security capabilties of [Fabric Private Chaincode (FPC)](https://github.com/hyperledger/fabric-private-chaincode).
+We describe the design of integration, including the user-facing changes. the required changes to the cc-tools and FPC, and limitations.
```
Comment (munapower) [inline README.md]: we describe the integration design ....

File: README.md [hunk 2]
```diff
@@ -1,83 +1,299 @@
...
-
-Simultaneously, CC Tools from Hyperledger Labs simplifies the chaincode lifecycle by providing tools to automate and streamline development, testing, and deployment. However, CC Tools lacks out-of-the-box support for FPC, leading to a fragmented development workflow that doesn't take full advantage of FPC's privacy capabilities while retaining CC Tools' development simplicity.
+This integration can't be done implicitly as both projects are not tailored to each other. CC-tools is created on the basis of using standard fapric networks so it doesn't handle encryption while communicating between the ledger and peers. FPC also was never tested with CC-tools packages as both are utilizing Fapric stub interface in their own stub wrapper
```
Comment (munapower) [inline README.md]: standard faBric networks .... it says faPric

File: README.md [hunk 3]
```diff
@@ -1,83 +1,299 @@
...
 
 The integration between FPC and CC Tools needs to occur at two distinct levels:
 
-Chaincode Level: The chaincode itself must be compatible with both the privacy features of FPC and the simplified lifecycle management offered by CC Tools. This requires modifications to the chaincode lifecycle, ensuring that it can be deployed securely using FPC, while retaining the flexibility of CC Tools for development and deployment.
+Chaincode Level: The chaincode itself will be written using CC Tools to simplify it's complexity and must be compatible with the privacy features of FPC. This requires modifications to the chaincode lifecycle, ensuring that it can be deployed securely using FPC, while retaining the flexibility of CC Tools for development.
+Note: Adhering to the security requirements by FPC adds more complexity to the chaincode deployment process than with standard fapric
```
Comment (munapower) [inline README.md]: Adhering to FPC's security requirements adds more ....

File: README.md [hunk 4]
```diff
@@ -1,83 +1,299 @@
...
 
-API Server (Client) Level: The API server (which is the client) that communicates with the peers must be adapted to handle the secure interaction between the FPC chaincode and the Fabric network. This includes managing secure communication with the TEE and ensuring that FPC’s privacy guarantees are upheld during transaction processing, while also utilizing CC Tools for chaincode operations.
+Client-Side Level: The client that communicates with the peers must be adapted to handle the secure interaction between the FPC chaincode and the Fabric network. This includes managing secure communication with the TEE and ensuring that FPC's privacy guarantees are upheld during transaction processing, while also utilizing CC Tools for chaincode operations. (Marcus: Does CC Tools actually provide any tooling for application integration, or is the api-server a testing tool?)
 
 Addressing these challenges is critical to enabling the seamless development and deployment of private chaincode on Hyperledger Fabric, ensuring both security and usability.
+Note: cc-tools is implementing functionalities yet not supported by FPC code
```
Comment (munapower) [inline README.md]: the following list comprises the functionality supported by cc-tools but not yet by FPC

File: README.md [hunk 5]
```diff
@@ -1,83 +1,299 @@
...
 
 ### ON THE CHAINCODE LEVEL
 
-In developing the chaincode for the integration of FPC and CC Tools, there are specific requirements to ensure that FPC’s privacy and security features are properly leveraged. First, both FPC and CC Tools wrap the shim.ChaincodeStub interface from Hyperledger Fabric, but they must be used in a specific order to avoid conflicts. To meet this requirement, the chaincode must be wrapped with the FPC stubwrapper before being passed to the CC Tools wrapper. This ensures that FPC’s privacy protections, such as secure execution within a trusted execution environment (TEE), are properly applied before the chaincode is handled by CC Tools for lifecycle management.
+In developing the chaincode for the integration of FPC and CC Tools, there are specific requirements to ensure that FPC's privacy and security features are properly leveraged. First, both FPC and CC Tools wrap the [shim.ChaincodeStub](https://github.com/hyperledger/fabric-chaincode-go/blob/acf92c9984733fb937fba943fbf7397d54368751/shim/interfaces.go#L28) interface from Hyperledger Fabric, but they must be used in a specific order to for it to function. The order is determined by few factors:
```
Comment (munapower) [inline README.md]: in a specific order for it to ....

Comment (munapower) [reply README.md]: This order is determined by...

File: README.md [hunk 6]
```diff
@@ -1,83 +1,299 @@
...
+   	f.rwset.AddWrite(key, value)
+
+   	// note that since we are not using the fabric proposal response  we can skip the putState call
+   	//return f.stub.PutState(key, value)
+   	return nil
+   }
+
+   ```
+
+   This was also tested practically be logging during invoking a transaction that uses the PutStat functionality
```
Comment (munapower) [inline README.md]: logging during the invocation of a transaction that uses the PutStatE

File: README.md [hunk 7]
```diff
@@ -1,83 +1,299 @@
...
+In developing the chaincode for the integration of FPC and CC Tools, there are specific requirements to ensure that FPC's privacy and security features are properly leveraged. First, both FPC and CC Tools wrap the [shim.ChaincodeStub](https://github.com/hyperledger/fabric-chaincode-go/blob/acf92c9984733fb937fba943fbf7397d54368751/shim/interfaces.go#L28) interface from Hyperledger Fabric, but they must be used in a specific order to for it to function. The order is determined by few factors:
+
+1. The order needed by the stub interface:
```
Comment (munapower) [inline README.md]: Stub interface order

File: README.md [hunk 8]
```diff
@@ -1,83 +1,299 @@
...
+
+   For this, It's better to inject the fpc stubwrapper in the middle between fabric stub and cc-tools stubwrapper
+2. The order needed by how the flow of the code work:
+   Since the cc-tools code is actually translated to normal fabric chaincode before communicating with the the ledger, the cc-tools code itself doesn't communicate with the ledger but performs some in-memory operations and then calls the same shim functionality from the fabric code (like explained above).
+
+   For FPC, it changes the way dealing with the ledger as it deals with decrypting the arguments before committing the transacion to the ledger and encrypting the response before sending back to the client.
+
+To meet this requirement, the chaincode must be wrapped with the FPC stubwrapper before being passed to the CC Tools wrapper. ![wrappingOrder](./wrappingOrder.png)
+
+Here's an example of how the enduser enables FPC for a CC-tools-based chaincode.
```
Comment (munapower) [inline README.md]: endorser

File: README.md [hunk 9]
```diff
@@ -1,83 +1,299 @@
...
+	server := &shim.ChaincodeServer{
+		CCID:     ccid,
+		Address:  address,
+		CC:       cc,
+		TLSProps: *tlsProps,
+	}
+	return server.Start()
+```
+
+Note: For this code to work, there are more changes required to be done in terms of packages, building, and the deployment process. We'll get to this in the [User Experience](#user-experience) section
```
Comment (munapower) [inline README.md]: we'll go INTO this in ...

File: README.md [hunk 10]
```diff
@@ -1,83 +1,299 @@
...
+4. The FPC Client SDK converts the attestation into publicly-verifiable evidence through the Intel Attestation Service (IAS), using the hardware manufacturer's root CA certificate as the root of trust (Steps 8-9).
+5. A `registerEnclave` transaction is invoked to validate and register the enclave credentials on the ledger, ensuring they match the chaincode definition, and making them available to channel members (Steps 10-12).
+
+![fpcFlow](./fpcFlow.png)
+
+For more details about this, refer to the FPC chaincode deployment process section in the FPC rfc document [here](https://github.com/hyperledger/fabric-rfcs/blob/main/text/0000-fabric-private-chaincode-1.0.md#deployment-process).
+
+### ON THE CLIENT-SIDE LEVEL
+
+Similarly, as cc-tools is a framework for developing the chaincode which is normally transalted to fabric chaincode (now also fabric private chaincode), the client transaction invokation flow is not affected by cc-tools but it must follow the FPC transaction flow for encryption and decryption operations to meet the security requirements asserted by FPC.
```
Comment (munapower) [inline README.md]: spell check this please

File: README.md [hunk 11]
```diff
@@ -1,83 +1,299 @@
...
 
-![wrappingOrder](./wrappingOrder.png)
+First, any chaincode that is supposed to be working with FPC need some dependencies to develop and build the chaincode like: Ego, Protocol Buffers, SGX PSW, etc...
 
-Here's an example of how we did it for the chaincode on cc-tools-demo with chaincode as a service
+For this we should follow this section in FPC repo to setup the development environment [here](https://github.com/hyperledger/fabric-private-chaincode/tree/main?tab=readme-ov-file#getting-started).
+
+###### 2. Develop the chaincode in cc-tools and fpc
+
+Fortunately, since the stubwrapper for both cc-tools and fpc are implementing the same interface, the conversion to an fpc chaincode can be done by plug-and-play. This means, the user should start by developing the chaincode using cc-tools and at the main loop where he passes the chaincode instance to the server to start it, he needs to wrap it with `fpc.NewPrivateChaincode`. For example, have a look at the cc-tools-demo chaincode below
```
Comment (munapower) [inline README.md]: change instead of referring to the user as HE to refer to them as THEY ... so they pass, the need, etc

**Review by munapower — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 2 above)
Comment (osamamagdy) [reply README.md]: Done for all occurrences

File: README.md (hunk 4 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 3 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 5 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 6 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 7 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 8 above)
Comment (osamamagdy) [reply README.md]: Instead of what?

File: README.md (hunk 9 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 11 above)
Comment (osamamagdy) [reply README.md]: Done

File: README.md (hunk 10 above)
Comment (osamamagdy) [reply README.md]: Done for the whole document

**Pull request merged**